compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java

```diff
@@ -3871,12 +3871,11 @@ private StackValue genCmpWithNull(KtExpression exp, IElementType opToken, @Nulla
         Type type = expressionType(exp);
         StackValue argument = pregeneratedExpr != null ? pregeneratedExpr : gen(exp);
 
-        if (kotlinType == null || TypeUtils.isNullableType(kotlinType)) {
+        if (kotlinType == null ||
+            !AsmUtil.isPrimitive(type) && (TypeUtils.isNullableType(kotlinType) || !InlineClassesUtilsKt.isInlineClassType(kotlinType))) {
```

Comment:
I've checked the behavior on a similar test, but with `String` replaced to `Int`, and kotlinc 1.3.50 returns "OK" for that too. So I think the `!AsmUtil.isPrimitive(type)` condition is not needed here. And similarly in the IR backend

Reply:
That's possible, but note that `StackValue.compareWithNull` would produce incorrect bytecode for primitive types and the same issue exists on the IR backend.

Your test shows that there is also an issue with this check, though... I'll take another look!